airgen/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+//! Compilation from powdr machines to AIRs
+
+use std::collections::BTreeMap;
+
+use ast::{
+    asm_analysis::{
+        AnalysisASMFile, InstructionDefinitionStatement, Machine, PilBlock, SubmachineDeclaration,
+    },
+    object::{Function, Instr, Link, LinkFrom, LinkTo, Location, Object, PILGraph},
+    parsed::{asm::InstructionBody, PilStatement},
+};
+
+const MAIN: &str = "Main";
+
+use number::FieldElement;
+
+pub fn compile<T: FieldElement>(input: AnalysisASMFile<T>) -> PILGraph<T> {
+    let main_location = Location::default().join("main");
+
+    // we start from the main machine
+    let main_ty = match input.machines.len() {
+        // if there is a single machine, treat it as main
+        1 => input.machines.keys().next().unwrap().clone(),
+        // otherwise, use the machine called `MAIN`
+        _ => {
+            assert!(input.machines.contains_key(MAIN));
+            MAIN.into()
+        }
+    };
+
+    // get a list of all machines to instantiate. The order does not matter.
+    let mut queue = vec![(main_location.clone(), main_ty.clone())];
+
+    let mut instances = vec![];
+
+    while let Some((location, ty)) = queue.pop() {
+        let machine = input.machines.get(&ty).unwrap();
+
+        queue.extend(machine.submachines.iter().map(|def| {
+            (
+                // get the absolute name for this submachine
+                location.clone().join(def.name.clone()),
+                // get its type
+                def.ty.clone(),
+            )
+        }));
+
+        instances.push((location, ty));
+    }
+
+    // visit the tree compiling the machines
+    let objects = instances
+        .into_iter()
+        .map(|(location, ty)| {
+            let object = ASMPILConverter::convert_machine(&location, &ty, &input);
+            (location, object)
+        })
+        .collect();
+
+    let main_ty = input.machines.get(&main_ty).unwrap();
```

Comment:
```suggestion
    let main_ty = input.machines.[&main_ty];
```
If you're always unwrapping this should also work (just preference nit, fine with as is too)